### https://ocw.mit.edu/courses/sloan-school-of-management/15-071-the-analytics-edge-spring-2017/linear-regression/assignment-2/reading-test-scores

# Problem 1.1 - Dataset size
#### Load the training and testing sets using the read.csv() function, and save them as variables with the names pisaTrain and pisaTest.
#### How many students are there in the training set?

In [1]:
pisaTrain  <- read.csv('../data//pisa2009train.csv') 
pisaTest <- read.csv('../data/pisa2009test.csv')
str(pisaTrain)

'data.frame':	3663 obs. of  24 variables:
 $ grade                : int  11 11 9 10 10 10 10 10 9 10 ...
 $ male                 : int  1 1 1 0 1 1 0 0 0 1 ...
 $ raceeth              : Factor w/ 7 levels "American Indian/Alaska Native",..: NA 7 7 3 4 3 2 7 7 5 ...
 $ preschool            : int  NA 0 1 1 1 1 0 1 1 1 ...
 $ expectBachelors      : int  0 0 1 1 0 1 1 1 0 1 ...
 $ motherHS             : int  NA 1 1 0 1 NA 1 1 1 1 ...
 $ motherBachelors      : int  NA 1 1 0 0 NA 0 0 NA 1 ...
 $ motherWork           : int  1 1 1 1 1 1 1 0 1 1 ...
 $ fatherHS             : int  NA 1 1 1 1 1 NA 1 0 0 ...
 $ fatherBachelors      : int  NA 0 NA 0 0 0 NA 0 NA 0 ...
 $ fatherWork           : int  1 1 1 1 0 1 NA 1 1 1 ...
 $ selfBornUS           : int  1 1 1 1 1 1 0 1 1 1 ...
 $ motherBornUS         : int  0 1 1 1 1 1 1 1 1 1 ...
 $ fatherBornUS         : int  0 1 1 1 0 1 NA 1 1 1 ...
 $ englishAtHome        : int  0 1 1 1 1 1 1 1 1 1 ...
 $ computerForSchoolwork: int  1 1 1 1 1 1 1 1 1 1 ...
 $ re

In [2]:
nrow(pisaTrain)

[1] 3663

# Problem 1.2 - Summarizing the dataset
#### Using tapply() on pisaTrain, what is the average reading test score of males?

In [3]:
tapply(pisaTrain$readingScore, pisaTrain$male, mean, na.rm = T)

0        1 
512.9406 483.5325

# Problem 1.3 - Locating missing values
#### Which variables are missing data in at least one observation in the training set? Select all that apply
<font color='red'>grade</font><br>
<font color='red'>male</font><br>
<font color='red'>raceeth</font> # Correct<br>
<font color='red'>preschool</font> # Correct<br>
<font color='red'>expectBachelors</font> # Correct<br>
<font color='red'>motherHS</font> # Correct<br>
<font color='red'>motherBachelors</font> # Correct<br>
<font color='red'>motherWork</font> # Correct<br>
<font color='red'>fatherHS</font> # Correct<br>
<font color='red'>fatherBachelors</font> # Correct<br>
<font color='red'>fatherWork</font> # Correct<br>
<font color='red'>selfBornUS</font> # Correct<br>
<font color='red'>motherBornUS</font> # Correct<br>
<font color='red'>fatherBornUS</font> # Correct<br>
<font color='red'>englishAtHome</font> # Correct<br>
<font color='red'>computerForSchoolwork</font> # Correct<br>
<font color='red'>read30MinsADay</font> # Correct<br>
<font color='red'>minutesPerWeekEnglish</font> # Correct<br>
<font color='red'>studentsInEnglish</font> # Correct<br>
<font color='red'>schoolHasLibrary</font> # Correct<br>
<font color='red'>publicSchool</font><br>
<font color='red'>urban</font><br>
<font color='red'>schoolSize</font> # Correct <br>
<font color='red'>readingScore</font><br>

In [4]:
sort(colnames(pisaTrain)[colSums(is.na(pisaTrain)) > 0])

[1] "computerForSchoolwork" "englishAtHome"         "expectBachelors"      
 [4] "fatherBachelors"       "fatherBornUS"          "fatherHS"             
 [7] "fatherWork"            "minutesPerWeekEnglish" "motherBachelors"      
[10] "motherBornUS"          "motherHS"              "motherWork"           
[13] "preschool"             "raceeth"               "read30MinsADay"       
[16] "schoolHasLibrary"      "schoolSize"            "selfBornUS"           
[19] "studentsInEnglish"

# Problem 1.4 - Removing missing values
### Linear regression discards observations with missing data, so we will remove all such observations from the training and testing sets. Later in the course, we will learn about imputation, which deals with missing data by filling in missing values with plausible information.
### Type the following commands into your R console to remove observations with any missing value from pisaTrain and pisaTest:
### pisaTrain = na.omit(pisaTrain)
### pisaTest = na.omit(pisaTest)
### How many observations are now in the training set?

In [5]:
pisaTrain = na.omit(pisaTrain)
pisaTest = na.omit(pisaTest)
nrow(pisaTrain)

[1] 2414

In [6]:
nrow(pisaTest)

[1] 990

# Problem 2.1 - Factor variables
### Factor variables are variables that take on a discrete set of values, like the "Region" variable in the WHO dataset from the second lecture of Unit 1. This is an unordered factor because there isn't any natural ordering between the levels. An ordered factor has a natural ordering between the levels (an example would be the classifications "large," "medium," and "small").
### Which of the following variables is an unordered factor with at least 3 levels? (Select all that apply.)
<font color='red'>grade</font><br>
<font color='red'>male</font><br>
<font color='red'>raceeth</font> # Correct<br>

<font color='red'>grade</font> # Correct <br>
<font color='red'>male</font><br>
<font color='red'>raceeth</font><br>

# Problem 2.2 - Unordered factors in regression models
#### To include unordered factors in a linear regression model, we define one level as the "reference level" and add a binary variable for each of the remaining levels. In this way, a factor with n levels is replaced by n-1 binary variables. The reference level is typically selected to be the most frequently occurring level in the dataset.
#### As an example, consider the unordered factor variable "color", with levels "red", "green", and "blue". If "green" were the reference level, then we would add binary variables "colorred" and "colorblue" to a linear regression problem. All red examples would have colorred=1 and colorblue=0. All blue examples would have colorred=0 and colorblue=1. All green examples would have colorred=0 and colorblue=0.
#### Now, consider the variable "raceeth" in our problem, which has levels "American Indian/Alaska Native", "Asian", "Black", "Hispanic", "More than one race", "Native Hawaiian/Other Pacific Islander", and "White". Because it is the most common in our population, we will select White as the reference level.
#### Which binary variables will be included in the regression model? (Select all that apply.)
<font color='red'>raceethAmerican Indian/Alaska Native</font> # Correct <br>
<font color='red'>raceethAsian</font> # Correct <br>
<font color='red'>raceethBlack</font> # Correct <br>
<font color='red'>raceethHispanic</font> # Correct <br>
<font color='red'>raceethMore than one race</font> # Correct <br>
<font color='red'>raceethNative Hawaiian/Other Pacific Islander</font> # Correct <br>
<font color='red'>raceethWhite</font><br>

In [7]:
table(pisaTrain$raceeth)


         American Indian/Alaska Native                                  Asian 
                                    20                                     95 
                                 Black                               Hispanic 
                                   228                                    500 
                    More than one race Native Hawaiian/Other Pacific Islander 
                                    81                                     20 
                                 White 
                                  1470 

# Problem 2.3 - Example unordered factors
### Consider again adding our unordered factor race to the regression model with reference level "White".
### For a student who is Asian, which binary variables would be set to 0? All remaining variables will be set to 1. (Select all that apply.)

<font color='red'>raceethAmerican Indian/Alaska Native</font> # Correct <br>
<font color='red'>raceethAsian</font><br>
<font color='red'>raceethBlack</font> # Correct <br>
<font color='red'>raceethHispanic</font> # Correct <br>
<font color='red'>raceethMore than one race</font> # Correct <br>
<font color='red'>raceethNative Hawaiian/Other Pacific Islander</font> # Correct <br>

### For a student who is white, which binary variables would be set to 0? All remaining variables will be set to 1. (Select all that apply.)

<font color='red'>raceethAmerican Indian/Alaska Native</font> # Correct <br>
<font color='red'>raceethAsian</font> # Correct <br>
<font color='red'>raceethBlack</font> # Correct <br>
<font color='red'>raceethHispanic</font> # Correct <br>
<font color='red'>raceethMore than one race</font> # Correct <br>
<font color='red'>raceethNative Hawaiian/Other Pacific Islander</font> # Correct <br>

# Problem 3.1 - Building a model
#### Because the race variable takes on text values, it was loaded as a factor variable when we read in the dataset with read.csv() -- you can see this when you run str(pisaTrain) or str(pisaTest). However, by default R selects the first level alphabetically ("American Indian/Alaska Native") as the reference level of our factor instead of the most common level ("White"). Set the reference level of the factor by typing the following two lines in your R console:
#### pisaTrain\\$raceeth = relevel(pisaTrain\\$raceeth, "White")
#### pisaTest\\$raceeth = relevel(pisaTest\\$raceeth, "White")
#### Now, build a linear regression model (call it lmScore) using the training set to predict readingScore using all the remaining variables.
#### It would be time-consuming to type all the variables, but R provides the shorthand notation "readingScore ~ ." to mean "predict readingScore using all the other variables in the data frame." The period is used to replace listing out all of the independent variables. As an example, if your dependent variable is called "Y", your independent variables are called "X1", "X2", and "X3", and your training data set is called "Train", instead of the regular notation:
#### LinReg = lm(Y ~ X1 + X2 + X3, data = Train)
#### You would use the following command to build your model:
#### LinReg = lm(Y ~ ., data = Train)
#### What is the Multiple R-squared value of lmScore on the training set?

In [8]:
pisaTrain$raceeth = relevel(pisaTrain$raceeth, "White")
pisaTest$raceeth = relevel(pisaTest$raceeth, "White")

lmScore  <- lm(readingScore ~ ., data=pisaTrain) # instead of type all variables names you can just type (.)
summary(lmScore)


Call:
lm(formula = readingScore ~ ., data = pisaTrain)

Residuals:
    Min      1Q  Median      3Q     Max 
-247.44  -48.86    1.86   49.77  217.18 

Coefficients:
                                                Estimate Std. Error t value
(Intercept)                                   143.766333  33.841226   4.248
grade                                          29.542707   2.937399  10.057
male                                          -14.521653   3.155926  -4.601
raceethAmerican Indian/Alaska Native          -67.277327  16.786935  -4.008
raceethAsian                                   -4.110325   9.220071  -0.446
raceethBlack                                  -67.012347   5.460883 -12.271
raceethHispanic                               -38.975486   5.177743  -7.528
raceethMore than one race                     -16.922522   8.496268  -1.992
raceethNative Hawaiian/Other Pacific Islander  -5.101601  17.005696  -0.300
preschool                                      -4.463670   3.486055  -1.280

### <font color='red'>Note</font> that this $R^2$ is lower than the ones for the models we saw in the lectures and recitation. This does not necessarily imply that the model is of poor quality. More often than not, it simply means that the prediction problem at hand (predicting a student's test score based on demographic and school-related variables) is more difficult than other prediction problems (like predicting a team's number of wins from their runs scored and allowed, or predicting the quality of wine from weather conditions).

# Problem 3.2 - Computing the root-mean squared error of the model
#### What is the training-set root-mean squared error (RMSE) of lmScore?

In [9]:
SSE  <- sum((lmScore$residuals)^2)
(SST  <- sqrt(SSE/nrow(pisaTrain)))

[1] 73.36555

# Problem 3.3 - Comparing predictions for similar students
#### Consider two students A and B. They have all variable values the same, except that student A is in grade 11 and student B is in grade 9. What is the predicted reading score of student A minus the predicted reading score of student B?
<font color='red'>-59.09</font><br>
<font color='red'>-29.54</font><br>
<font color='red'>0</font><br>
<font color='red'>29.54</font><br>
<font color='red'>59.09</font><br>
<font color='red'>The difference cannot be determined without more information about the two students</font><br>

In [10]:
# becouse Estimate of variable (grade) is 29.542707 so we compute:
(11 - 9) * 29.542707 

[1] 59.08541

# Problem 3.4 - Interpreting model coefficients
#### What is the meaning of the coefficient associated with variable raceethAsian?
<font color='red'>Predicted average reading score of an Asian student</font><br>
<font color='red'>Difference between the average reading score of an Asian student and the average reading score of a white student</font><br>
<font color='red'>Difference between the average reading score of an Asian student and the average reading score of all the students in the dataset</font><br>
<font color='red'>Predicted difference in the reading score between an Asian student and a white student who is otherwise identical</font> # Correct <br>

# Problem 3.5 - Identifying variables lacking statistical significance
#### Based on the significance codes, which variables are candidates for removal from the model? Select all that apply. (We'll assume that the factor variable raceeth should only be removed if none of its levels are significant.)
<font color='red'>grade</font><br>
<font color='red'>male</font><br>
<font color='red'>raceeth</font><br>
<font color='red'>preschool</font> # Correct <br>
<font color='red'>expectBachelors</font><br>
<font color='red'>motherHS</font> # Correct <br>
<font color='red'>motherBachelors</font><br>
<font color='red'>motherWork</font> # Correct <br>
<font color='red'>fatherHS</font> # Correct <br>
<font color='red'>fatherBachelors</font><br>
<font color='red'>fatherWork</font> # Correct <br>
<font color='red'>selfBornUS</font> # Correct <br>
<font color='red'>motherBornUS</font> # Correct <br>
<font color='red'>fatherBornUS</font> # Correct <br>
<font color='red'>englishAtHome</font> # Correct <br>
<font color='red'>computerForSchoolwork</font><br>
<font color='red'>read30MinsADay</font><br>
<font color='red'>minutesPerWeekEnglish</font> # Correct <br>
<font color='red'>studentsInEnglish</font> # Correct <br>
<font color='red'>schoolHasLibrary</font> # Correct <br>
<font color='red'>publicSchool</font><br>
<font color='red'>urban</font> # Correct <br>
<font color='red'>schoolSize</font><br>

# Problem 4.1 - Predicting on unseen data
### Using the "predict" function and supplying the "newdata" argument, use the lmScore model to predict the reading scores of students in pisaTest. Call this vector of predictions "predTest". Do not change the variables in the model (for example, do not remove variables that we found were not significant in the previous part of this problem). Use the summary function to describe the test set predictions.
### What is the range between the maximum and minimum predicted reading score on the test set?

In [11]:
predTest <- predict(lmScore, newdata = pisaTest)
max(predTest) - min(predTest)

[1] 284.4683

# Problem 4.2 - Test set SSE and RMSE
### What is the sum of squared errors (SSE) of lmScore on the testing set?

In [12]:
(SSE <- sum((pisaTest$readingScore - predTest)^2))

[1] 5762082

### What is the root-mean squared error (RMSE) of lmScore on the testing set?

In [13]:
(RMSE  <- sqrt(SSE / nrow(pisaTest)))

[1] 76.29079

# Problem 4.3 - Baseline prediction and test-set SSE
### What is the predicted test score used in the baseline model? Remember to compute this value using the training set and not the test set.

In [14]:
mean(pisaTrain$readingScore)

[1] 517.9629

### What is the sum of squared errors of the baseline model on the testing set? HINT: We call the sum of squared errors for the baseline model the total sum of squares (SST).

In [15]:
sum((pisaTest$readingScore - 517.962887323944)^2)

[1] 7802354

# Problem 4.4 - Test-set R-squared
### What is the test-set R-squared value of lmScore?

In [16]:
SSE <- sum((pisaTest$readingScore - predTest)^2)
SST <- sum((pisaTest$readingScore - mean(pisaTrain$readingScore))^2)
(R2 <- 1 - SSE / SST)

[1] 0.2614944